credentials/tls.go

```diff
@@ -112,6 +116,22 @@ func (c *tlsCreds) ClientHandshake(ctx context.Context, authority string, rawCon
 		conn.Close()
 		return nil, nil, ctx.Err()
 	}
+
+	// The negotiated protocol can be either of the following:
+	// 1. h2: When the server supports ALPN. Only HTTP/2 can be negotiated since
+	//    it is the only protocol advertised by the client during the handshake.
+	//    The tls library ensures that the server chooses a protocol advertised
+	//    by the client.
+	// 2. "" (empty string): If the server doesn't support ALPN. ALPN is a requirement
+	//    for using HTTP/2 over TLS. We can terminate the connection immediately.
+	np := conn.ConnectionState().NegotiatedProtocol
+	if np == "" {
```

Comment:
In addition to checking if np is empty (""), shouldn't we also validate the value of np against a list of unsupported protocol like "h2c" (as called out in the [issue](https://github.com/grpc/grpc-go/issues/434))? Alternatively, we can flip the check to np != "h2" if that's the only supported protocol or can maintain a list of supported protocols.

Reply:
I checked all the places where a new tls.Config is being created and we're setting the NextProtos field to `{"h2"}` everywhere. However, we're also calling the `AppendH2ToNextProtos` method (https://github.com/grpc/grpc-go/blob/master/internal/credentials/util.go#L28-L37) in all the constructions of the TLS `TransportCredentials`. It seems that gRPC code is written to allow addition of more protocols in the future.

We are using the crypto/tls library to do the TLS handshake on the client side. If we set only "h2" in the `tls.Config.NextProtos` slice, [this function](https://github.com/golang/go/blob/master/src/crypto/tls/handshake_client.go#L828-L845) ensures that the application protocol selected by the server is present in the list supported by the client. If not, the tls library fails the handshake. So the re-validation of protocol selected by the server in gRPC code seems redundant.

Reply:
Thanks! This has been discussed in the issue #434 . I think I might have missed it. Appreciate summarizing it here.